### ISSUE TITLE
.bind(foo='foo')  does not work for Retrievers

### ISSUE THREAD

**Issue by hugoJuhel** [langchain-ai/langchain#16547]
### Checked other resources

- [X] I added a very descriptive title to this issue.
- [X] I searched the LangChain documentation with the integrated search.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I am sure that this is a bug in LangChain rather than my code.

### Example Code

```python
from langchain_core.retrievers import BaseRetriever
from langchain_core.callbacks import CallbackManagerForRetrieverRun
from langchain_core.documents import Document
from typing import List


class CustomRetriever(BaseRetriever):

    # the signature is slightly modified with the addition of `spam`
    def _get_relevant_documents(
        self, query: str, *, spam, run_manager: CallbackManagerForRetrieverRun
    ) -> List[Document]:
        return [Document(page_content=query['query'])]


retriever = CustomRetriever()

# Binding a spam attribute, 
spam = {'query': RunnablePassthrough()} | retriever.bind(spam=2)
spam.invoke("bar")
```

### Error Message and Stack Trace (if applicable)

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[33], [line 17](vscode-notebook-cell:?execution_count=33&line=17)
     [14](vscode-notebook-cell:?execution_count=33&line=14) retriever = CustomRetriever()
     [16](vscode-notebook-cell:?execution_count=33&line=16) spam = {'query': RunnablePassthrough()} | retriever.bind(spam=2)
---> [17](vscode-notebook-cell:?execution_count=33&line=17) spam.invoke("bar")

TypeError: BaseRetriever.invoke() got an unexpected keyword argument 'spam'

### Description

# Details
`.bind()` does not seem to be working well with `Retrievers`.
The `BaseRetriever.invoke()` does not support variadics arguments. Runtime binding on a retriever raises a `TypeError`.
A possible workaround is to use a `RunnableLambda` wrapping the `retriever.get_relevant_documents` and dispatching the bounded arguments to a `get_relevant_documents. `.

# How to reproduce
Instantiate a `CustomRetriever` as per [Langchain's documentation](https://python.langchain.com/docs/modules/data_connection/retrievers/) and `bind` an argument to it.

# What to expect

`bind` should dispatch bounded argument to `retrievers` too.

### System Info

System Information
------------------
> OS:  Linux
> OS Version:  #1 SMP Fri Apr 2 22:23:49 UTC 2021
> Python Version:  3.11.6 (main, Nov  1 2023, 14:10:18) [GCC 9.4.0]

Package Information
-------------------
> langchain_core: 0.1.15
> langchain: 0.1.0
> langchain_community: 0.0.12
> langchain_openai: 0.0.2.post1

Packages not installed (Not Necessarily a Problem)
--------------------------------------------------
The following packages were not found:

> langgraph
> langserve

**Comment by baskaryan** [comment]
You'd want to overwrite the `invoke` function as well to pass through kwargs:

```python
from langchain_core.retrievers import BaseRetriever
from langchain_core.callbacks import CallbackManagerForRetrieverRun
from langchain_core.documents import Document
from langchain_core.runnables import RunnableConfig, RunnablePassthrough, ensure_config
from typing import Any, List, Optional


class CustomRetriever(BaseRetriever):

    # the signature is slightly modified with the addition of `spam`
    def _get_relevant_documents(
        self, query: str, *, spam, run_manager: CallbackManagerForRetrieverRun
    ) -> List[Document]:
        return [Document(page_content=query['query'])]

    def invoke(
        self, input: str, config: Optional[RunnableConfig] = None, **kwargs: Any
    ) -> List[Document]:
        config = ensure_config(config)
        return self.get_relevant_documents(
            input,
            callbacks=config.get("callbacks"),
            tags=config.get("tags"),
            metadata=config.get("metadata"),
            run_name=config.get("run_name"),
           **kwargs
        )

retriever = CustomRetriever()

# Binding a spam attribute, 
spam = {'query': RunnablePassthrough()} | retriever.bind(spam=2)
spam.invoke("bar")
```

@nfcampos should we update BaseRetriever.invoke to do this be default?

**Comment by hugoJuhel** [comment]
@baskaryan 

Hi. Thanks for your help. It might be silly but shouldn't `_aget_relevant_documents` also accept variadic arguments, for `ainvoke` to support binding ?

**Comment by baskaryan** [comment]
> @baskaryan
> 
> Hi. Thanks for your help. It might be silly but shouldn't `_aget_relevant_documents` also accept variadic arguments, for `ainvoke` to support binding ?

Yep! Both should be fixed in #16551, which will be out in next langchain-core release
